cpp/src/arrow/engine/substrait/extension_set.h

```diff
@@ -95,6 +96,17 @@ class ARROW_ENGINE_EXPORT ExtensionIdRegistry {
   virtual Status CanRegisterFunction(Id,
                                      const std::string& arrow_function_name) const = 0;
   virtual Status RegisterFunction(Id, std::string arrow_function_name) = 0;
+
+  /// \brief Add a symbol external to the plan yet used in an Id.
+  ///
+  /// This ensures the symbol, which is only viewed but not held by the Id, lives while
+  /// the extension set does. Symbols appearing in the Substrait plan are already held.
+  const std::string& AddExternalSymbol(const std::string& symbol) {
+    return *external_symbols.insert(symbol).first;
+  }
```

Comment:
I think we should have...

```
// Registers the function and takes ownership of uri, name
RegisterFunction(std::string uri, std::string name, std::string arrow_function_name) = 0;
// Registers the function without taking ownership of id
RegisterFunction(Id id, std::string arrow_function_name) = 0;
```

Then `external_symbols` can be an implementation detail.

Reply:
I'll fix.